appengine/genetics/js/cage.js

```diff
@@ -188,55 +215,92 @@ Genetics.Cage.reset = function() {
  * @param {string|!Function} code Player's code, or generator.
  */
 Genetics.Cage.addPlayer = function(playerName, code) {
-  // Assign a unique ID to the player and add to the game.
-  var id = Genetics.Cage.nextAvailablePlayerId_++;
-  Genetics.Cage.players[id] = [playerName, code];
+  Genetics.Cage.players.push({name: playerName, code: code});
 };
 
 /**
  * Start the Cage simulation. Players should be already added.
- * @param {Function} doneCallback Function to call when game ends.
- */
-Genetics.Cage.start = function(doneCallback) {
-  Genetics.Cage.doneCallback_ = doneCallback;
-  // Create mice for each player.
-  for (var playerId in Genetics.Cage.players) {
-    for (var i = 0; i < Genetics.Cage.START_MICE_PER_PLAYER; i++) {
-      var mouseSex = i % 2 == 0 ? Genetics.Mouse.Sex.MALE :
-          Genetics.Mouse.Sex.FEMALE;
-      var mouseId = Genetics.Cage.nextAvailableMouseId_++;
-      var mouse = new Genetics.Mouse(mouseId, mouseSex, null, null, playerId);
-      Genetics.Cage.born(mouse);
-      new Genetics.Cage.Event('ADD', goog.object.clone(mouse),
-          Genetics.Cage.players[playerId][0]).addToQueue();
+ * @param {!Function} checkForEnd Function to call when game ends.
+ * that should be skipped in simulateLife.
+ */
+Genetics.Cage.start = function(checkForEnd) {
+  Genetics.Cage.checkForEnd = checkForEnd;
+  Genetics.Cage.stopped_ = false;
+  Genetics.Cage.nextAvailableMouseId_ = Genetics.Cage.nextRoundMice_.length;
+  for (var playerId = 0, player; player = Genetics.Cage.players[playerId];
+      playerId++) {
+    if (goog.isFunction(player.code)) {
+      // Cache code if player code is generator.
+      player.cachedCode = Genetics.Cage.players[1].code + '\n' + player.code();
```

Comment:
code can either be a string or a function, so getCode wouldn't quite work for that. T